### ISSUE TITLE
Include MPI_Scan and MPI_Exscan in All-to-All classification

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#134]
**Originally by _moody20_ on 2009-02-13 13:29:26 -0600**

---
# Include MPI_Scan and MPI_Exscan in All-to-All classificationAuthor: Adam Moody
## Description

Each process contributes data and receives data in MPI_Scan and (with the exception of rank 0) MPI_Exscan.  This proposal moves them from the Other classification to the All-to-All classification, and along with Ticket #89 empties out the Other category.

As written, this proposal currently depends on Ticket #89 passing.
## Proposed Solution

Move "MPI_SCAN, MPI_EXSCAN" bullet from p 133, line 23 to p 133, line 11, just after the reductions in the All-to-All category.

Remove p 133, line 22:
  Other  Collective operations that do not fit into one of the above categories.

Remove p 133, line 27-28:
  The data movement patterns of MPI_SCAN and MPI_EXSCAN do not fit this taxonomy.
## Impact on Implementations

None: only text changes.
## Impact on Applications / Users

None: only text changes.
## Alternative Solutions

N/A
## Entry for the Change Log

None: only text changes.

**Comment by mpiforumbot** [comment]
**Originally by _bronis_ on 2009-02-13 13:47:58 -0600**

---

Looks fine.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-02-28 22:05:49 -0600**

---

I disagree to this proposal. The phrasing with "result" instead of just "data" (which could be anything) seems important to me. I understand the concern that there is no single result, but I'd like to keep the notion of a "result". One could say "all processes receive (part of) the result" or even better "receive a subset of the result" (not necessarily a proper subset). However, replacing this with "data" looses the notion that this is actually an "all-to-all" communication, i.e., each process receives some contribution from all other processes. This ("all-to-all") is not true for MPI_{EX}SCAN, thus, those calls are excluded from this taxonomy. They would not fit an "all-to-all" classification, more like a "some-to-some", but that's hard to specify. Thus, I'd propose to leave the taxonomy unchanged.

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-03-04 04:35:16 -0600**

---

I also disagree with the proposal, and see no added value in forcing the special Scan/Exscan
functions into a classification where they don't really fit. Also barrier in my opinion should
be treated as special in the classification, thus I think I disagree with ticket #89

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-03-05 14:23:27 -0600**

---

Actually, MPI_Scan does indeed fit the alltoall category -- all processes contribute to the result, and all processes receive the result.

MPI_Exscan technically does not -- all processes contribute, all but rank 0 receives.  However, having said that, alltoallv can also produce the same pattern as exscan -- just set the recvcounts for rank 0 to be 0.  So in some sense, we are already forcing the *v and *w collectives into categories that maybe they don't quite fit.  But that's ok, the point of this section is not to be technically accurate, but to make these broad categorizations in order to make help convey a concept.

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-03-23 05:46:30 -0500**

---

I still think it contrived to put Scan/Exscan into the alltoall category. And I think the usefulness of such a classification is much diminished by force-fitting into a general category. Same comment on Ticket #89.

Jesper

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2009-04-02 14:04:30 -0500**

---

The proposal itself is fine.  

I will vote against it.  :-)
- I can see moving MPI_SCAN to all-to-all; all processes do contribute to and receive the result.  Hence, it fits the definition provided on p133:6.
- MPI_EXSCAN does not fit in this definition because rank 0 does not receive any result.  It still belongs in "other".  I will not vote for something that is clearly wrong!  ;-)
- I agree with Torsten's rationale about "data" and "result".

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-04-06 15:39:17 -0500**

---

Ok, I'm fine with keeping "result" instead of specifying "data".  However, I still think it's useful to add these in the categories.  These categories (so far) are not being used to make general categorical definitions for MPI -- they are defined simply to illustrate the data movement in intercommunicators.  For this purpose, I think it's very helpful to include scan and exscan (assuming the ticket to extend scan and exscan for intercommunicators passes).

Anyway, my argument is largely dependent on whether Ticket #133 for scan/exscan passes, so I'll place this on hold.

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-04-07 14:01:31 -0500**

---

Keeping "result" instead of newly proposed "data".

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-04-08 09:28:29 -0500**

---

Since rank 0 does not receive any data from other processes (only itself), it technically not an All-to-All.  I'm withdrawing the ticket.
